### PR TITLE
Shoaib ansari 1162

### PR DESCRIPTION
**Pull request by MohdShoaibClassic**
Done.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -9,10 +9,12 @@
...
+    "@reduxjs/toolkit": "^1.9.2",
     "react": "^18.2.0",
     "react-dom": "^18.2.0",
-    "react-redux": "^8.0.5"
+    "react-redux": "^8.0.5",
+    "react-router-dom": "^6.8.0",
+    "react-toastify": "4.1"
   },
   "devDependencies": {
     "@types/react": "^18.0.26",
```
Comment (atmeshwar-classic) [inline package.json:20]: We already have yarn.lock so this file is not needed.
It is a bad practice to keep two lockfiles in the same project.

Please use yarn to install dependencies.

Comment (MohdShoaibClassic) [reply package.json:20]: Okay fine. I will take care these things onwards.

File: src/App.css [hunk 2]
```diff
@@ -1,41 +1,294 @@
-#root {
-  max-width: 1280px;
-  margin: 0 auto;
-  padding: 2rem;
+@import url("https://fonts.googleapis.com/css2?family=Nunito:wght@200;400;700&display=swap");
```
Comment (atmeshwar-classic) [inline src/App.css:1]: We should not write all of our CSS in one file. Please extract the component specific styles to their respective files.

File: src/App.tsx [hunk 3]
```diff
@@ -1,32 +1,23 @@
...
-          count is {count}
-        </button>
-        <p>
-          Edit <code>src/App.tsx</code> and save to test HMR
-        </p>
-      </div>
-      <p className="read-the-docs">
-        Click on the Vite and React logos to learn more
-      </p>
+      <BrowserRouter>
```
Comment (atmeshwar-classic) [inline src/App.tsx:10]: We should always write our routes in its separate file and then import them here. Like:
src/routes/ routes.ts

We should always structure our app regardless of its size.

File: src/domain/books/BooksPage.tsx [hunk 4]
```diff
@@ -1,7 +1,32 @@
...
+                <BookCard {...book} />
+              )
+              )}
+          </div>
+        </>
+      ) : loading === "loading" ? (
+        <p>Loading...</p>
+      ) : (
+        <p>Unexpected error occured...</p>
+      )}
```
Comment (atmeshwar-classic) [inline src/domain/books/BooksPage.tsx]: 1) Toast container should be at the root, ideally inside main.tsx.

2) Avoid nested ternaries. Instead, extract it to a function and use if with early return
`const renderContent = () => {
  if (loading === 'loading') {
   return <></>
  }

 if (loading === 'error') {
  return <></>
 }
 
 return <><</>
}`

3 books is not undefined or null so need to put a guard in front of it.

File: src/domain/cart/CartPage.tsx [hunk 5]
```diff
@@ -1,5 +1,74 @@
...
+const CartPage = (): JSX.Element => {
+  const cart = useAppSelector(cartSelector);
+  const dispatch = useAppDispatch();
+
+  const handleRemoveFromCart = (cartItem: CartItem) => {
+    dispatch(removeFromCart(cartItem));
+  };
+
+  useEffect(() => {
+    dispatch(getTotals());
```
Comment (atmeshwar-classic) [inline src/domain/cart/CartPage.tsx]: getTotals() should not be in redux. Instead, put in a utils file and import it here.
Also, the useEffect here is redundant. You can call the fn where you need its value and if it's an expensive fn then use useMemo.

File: src/domain/cart/cartSlice.ts [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+                },
+                {
+                    total: 0,
+                    quantity: 0,
+                }
+            );
+            total = parseFloat(total.toFixed(2));
+            state.cartTotalQuantity = quantity;
+            state.cartTotalAmount = total;
+        },
```
Comment (atmeshwar-classic) [inline src/domain/cart/cartSlice.ts]: As stated above, it should not be here.

**Review by atmeshwar-classic — CHANGES_REQUESTED** [review]
Hi @MohdShoaibClassic , please look at my comments.

File: src/store/store.ts [hunk 7]
```diff
@@ -1,13 +1,17 @@
...
+  books: booksSliceReducer,
+  cartItems: cartSliceReducer
 });
 
 export const store = configureStore({
   reducer: combinedReducer
 })
 
+store.dispatch(getBooks());
+
```
Comment (atmeshwar-classic) [inline src/store/store.ts]: Do not dispatch this action here. We should call it inside BooksPage's useEffect.

File: src/utils/total.ts [hunk 8]
```diff
@@ -0,0 +1,4 @@
+import { CartState } from "../domain/cart/types";
+export const total = (cart:CartState) => {
```
Comment (atmeshwar-classic) [inline src/utils/total.ts]: No need to import the entire CartState here. We just need the cartItems.

File: src/routes/Routes.tsx [hunk 9]
```diff
@@ -1,12 +0,0 @@
...
-import CartPage from "../domain/cart/CartPage";
-const Routes = () => {
-    return (
-        <Switch>
-            <Route path="/cart" element={<CartPage />} />
-            <Route path="/" element={<BooksPage />} />
-        </Switch>
-    )
-}
-export default Routes
```
Comment (atmeshwar-classic) [inline src/routes/Routes.tsx]: This was the correct way to handle routes. Why did you delete it?

File: src/App.tsx [hunk 10]
```diff
@@ -1,15 +1,18 @@
...
+
   return (
     <div className="App">
       <BrowserRouter>
         <NavBar />
         <div className="content-container">
-          <Routes/>
+          <Routes>
+            {routes.map(({ path, Component }) => <Route path={path} element={<Component />} />)}
+          </Routes>
```
Comment (atmeshwar-classic) [inline src/App.tsx]: Please see the below comment

**Comment by atmeshwar-classic** [comment]
I am closing the PR as the assessment is complete.
Good job @MohdShoaibClassic . Keep it up!
